### PR TITLE
WIP Add support for transforming constructor names.

### LINKED ISSUE #110: Support constructor name transformations
We support `transformMemberNames`, for changing the names of case class members in their JSON representation, but don't yet support `transformConstructorNames` for ADTs. The behavior should be the same as that provided by [circe-generic-extras](https://github.com/circe/circe-generic-extras).

### PR DESCRIPTION
**Pull request by danxmoran**
(Eventually) resolves #110

Still WIP because there are some open questions, but the code that's there is ready for a look.

### CODE REVIEW

File: modules/annotations/shared/src/main/scala/io/circe/derivation/annotations/Configuration.scala [hunk 1]
```diff
@@ -18,12 +18,20 @@ sealed trait Configuration {
...
+  def transformConstructorNames: String => String
+
   def useDefaults: Boolean
 
   def discriminator: Option[String]
 
   protected def getA(transformMemberNames: String => String): Config
 
+  // TODO: Naming???
+  protected def getC(transformConstructorNames: String => String): Config
```
Comment (danxmoran) [inline modules/annotations/shared/src/main/scala/io/circe/derivation/annotations/Configuration.scala]: I wasn't sure if the existing `getA` meant something that I should try to mimic here. I chose `C` for constructor 😄

File: modules/annotations/shared/src/main/scala/io/circe/derivation/annotations/JsonCodec.scala [hunk 2]
```diff
@@ -10,6 +10,7 @@ class JsonCodec(
   def macroTransform(annottees: Any*): Any = macro GenericJsonCodecMacros.jsonCodecAnnotationMacro
 }
 
+// FIXME: Should these annotations also handle constructor transformations?
```
Comment (danxmoran) [inline modules/annotations/shared/src/main/scala/io/circe/derivation/annotations/JsonCodec.scala]: Biggest open question feature-wise, I think

File: modules/derivation/shared/src/main/scala/io/circe/derivation/DerivationMacros.scala [hunk 3]
```diff
@@ -351,8 +358,13 @@ class DerivationMacros(val c: blackbox.Context) extends ScalaVersionCompat {
           }
         }
       case _root_.scala.Some(typeFieldName) =>
-        _root_.io.circe.Decoder[_root_.java.lang.String].prepare(_.downField(typeFieldName)).flatMap {
-          case ..$discriminatorCases
+        new _root_.io.circe.Decoder[$tpe] {
+          // NOTE: This wrapper might look pointless, but leaving it out causes compilation to fail with
+          // an assertion error from the compiler. There's probably a better way to avoid the problem.
```
Comment (danxmoran) [inline modules/derivation/shared/src/main/scala/io/circe/derivation/DerivationMacros.scala]: This is the problem I mentioned in the ticket. I noticed that the compiler crash only happened in the `Decoder` case, not the `Codec`. The only difference I could see was that the `Codec` explicitly created a wrapper class around the generated `case` statements, so I tried doing that and it worked. This is my first time digging into Scala macros so I have no idea why the other way wouldn't work.

File: modules/annotations/shared/src/main/scala/io/circe/derivation/annotations/JsonCodec.scala (hunk 2 above)
Comment (travisbrown) [reply modules/annotations/shared/src/main/scala/io/circe/derivation/annotations/JsonCodec.scala]: I'd vote no.

File: modules/annotations/shared/src/main/scala/io/circe/derivation/annotations/Configuration.scala (hunk 1 above)
Comment (travisbrown) [reply modules/annotations/shared/src/main/scala/io/circe/derivation/annotations/Configuration.scala]: That was @aparo's name. I don't personally feel strongly about it, and given that they're protected and on a sealed trait I don't think we need to worry too much about the decision.

File: modules/derivation/shared/src/main/scala/io/circe/derivation/DerivationMacros.scala [hunk 4]
```diff
@@ -592,24 +602,27 @@ class DerivationMacros(val c: blackbox.Context) extends ScalaVersionCompat {
 
   // we materialize trait
   private[this] def materializeEncoderTraitImpl[T: c.WeakTypeTag](
-    transformMemberNames: Option[c.Expr[String => String]],
+    transformConstructorNames: Option[c.Expr[String => String]],
```
Comment (travisbrown) [inline modules/derivation/shared/src/main/scala/io/circe/derivation/DerivationMacros.scala]: Good catch that we didn't need the old argument here.

File: modules/derivation/shared/src/main/scala/io/circe/derivation/DerivationMacros.scala [hunk 5]
```diff
@@ -678,39 +692,40 @@ class DerivationMacros(val c: blackbox.Context) extends ScalaVersionCompat {
...
 
+    def transformName(name: String): Tree = transformConstructorNames.fold[Tree](q"$name")(f => q"$f($name)")
+
     val objectWrapperCases: List[Tree] = subclasses.map { s =>
-      val value =
-        Literal(Constant(nameOf(s)))
+      val value = transformName(nameOf(s))
 
-      cq"""$value => c.get($value)(_root_.io.circe.Decoder[${s.asType}]).asInstanceOf[_root_.io.circe.Decoder.Result[$tpe]]"""
+      cq"""nme if nme == $value => c.get(nme)(_root_.io.circe.Decoder[${s.asType}]).asInstanceOf[_root_.io.circe.Decoder.Result[$tpe]]"""
```
Comment (travisbrown) [inline modules/derivation/shared/src/main/scala/io/circe/derivation/DerivationMacros.scala]: Out of curiosity, was this change necessary?

Comment (danxmoran) [reply modules/derivation/shared/src/main/scala/io/circe/derivation/DerivationMacros.scala]: It was (as far as I could tell). Matching directly on `$value` caused compilation errors about function 1 not having an `unapply` method when `transformConstructorNames` wasn't `None`. The generated code looked like:
```scala
case (((x: String) => someFunc(x))("AdtBar")) => ???
case (((x: String) => someFunc(x))("AdtFoo")) => ???
```
Now they look like:
```scala
case nme if nme == (((x: String) => someFunc(x))("AdtBar")) => ???
case nme if nme == (((x: String) => someFunc(x))("AdtFoo")) => ???
```

Related question I was wondering about: is there any way to "lift" the transformation function into the macro to make it directly useable, so we can pre-compute the transformed strings?

File: modules/derivation/shared/src/main/scala/io/circe/derivation/DerivationMacros.scala [hunk 6]
```diff
@@ -738,10 +753,12 @@ class DerivationMacros(val c: blackbox.Context) extends ScalaVersionCompat {
...
       case _root_.scala.Some(typeFieldName) =>
         new _root_.io.circe.Codec.AsObject[$tpe] {
-          private[this] val decoder =
-            _root_.io.circe.Decoder[_root_.java.lang.String].prepare(_.downField(typeFieldName)).flatMap {
-              case ..$discriminatorCases
-            }
+          private[this] val decoder = new _root_.io.circe.Decoder[$tpe] {
+            def apply(c: _root_.io.circe.HCursor): _root_.io.circe.Decoder.Result[$tpe] =
+              c.get(typeFieldName)(_root_.io.circe.Decoder[_root_.java.lang.String]).flatMap {
+                case ..$discriminatorCases
```
Comment (danxmoran) [inline modules/derivation/shared/src/main/scala/io/circe/derivation/DerivationMacros.scala]: Other question I remembered: This code path didn't have a catch-all `case _ => ???` like the object-wrapper case did. Is that intentional? I don't think the generated cases are exhaustive.

Comment (travisbrown) [reply modules/derivation/shared/src/main/scala/io/circe/derivation/DerivationMacros.scala]: I don't think I was the original author, but we definitely want this to be exhaustive. We could do that in a separate PR, or here. If you can come up with a failing test in addition that would be amazing.

Comment (danxmoran) [reply modules/derivation/shared/src/main/scala/io/circe/derivation/DerivationMacros.scala]: I'm ok adding it + test here :+1:

**Comment by danxmoran** [comment]
Any tips for debugging the JS errors? The problem only seems to be in `deriveDecoder`, which is weird because the new code is exactly the same as in `deriveCodec`.

**Comment by travisbrown** [comment]
@danxmoran Does this branch have the 0.6.29 update? If so you might try rolling back to 0.6.28. It's a compiler crash, so I suspect it's not our fault. I can take a closer look later.

**Comment by danxmoran** [comment]
@travisbrown I hit the crash on both 0.6.28 and 0.6.29

**Comment by travisbrown** [comment]
These commits are included in #186. Thanks again, @danxmoran!
